### PR TITLE
ddl: Save reorg doneHandle regularly

### LINKED ISSUE #4986: DDL job always run.I can't kill it!
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
Look at Id 46.I want to add a index,but the ddl job always run,I can't kill it.
```
tidb> admin show ddl jobs;
+------------------------------------------------------------------------------------------------------------------------------+---------+
| JOBS                                                                                                                         | STATE   |
+------------------------------------------------------------------------------------------------------------------------------+---------+
| ID:46, Type:add index, State:running, SchemaState:write reorganization, SchemaID:1, TableID:33, RowCount:165059980, ArgLen:0 | running |
| ID:47, Type:drop index, State:none, SchemaState:none, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                           | none    |
| ID:48, Type:drop index, State:none, SchemaState:none, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                           | none    |
| ID:45, Type:drop column, State:synced, SchemaState:none, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                        | synced  |
| ID:44, Type:drop column, State:synced, SchemaState:none, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                        | synced  |
| ID:43, Type:drop column, State:synced, SchemaState:none, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                        | synced  |
| ID:42, Type:add column, State:synced, SchemaState:public, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                       | synced  |
| ID:41, Type:add column, State:synced, SchemaState:public, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                       | synced  |
| ID:40, Type:drop index, State:synced, SchemaState:none, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                         | synced  |
| ID:39, Type:add column, State:synced, SchemaState:public, SchemaID:1, TableID:33, RowCount:0, ArgLen:0                       | synced  |
| ID:38, Type:add index, State:synced, SchemaState:public, SchemaID:1, TableID:33, RowCount:100283953, ArgLen:0                | synced  |
| ID:37, Type:drop table, State:synced, SchemaState:none, SchemaID:1, TableID:35, RowCount:0, ArgLen:0                         | synced  |
| ID:36, Type:create table, State:synced, SchemaState:public, SchemaID:1, TableID:35, RowCount:0, ArgLen:0                     | synced  |
+------------------------------------------------------------------------------------------------------------------------------+---------+
```

2. What did you expect to see?
I want the id 46 job success and stop.

3. What did you see instead?
The total number of my table is 100283953 , but the Id 46 job  show ``RowCount:165059980.``  and the number still increase .

I want to kill it. But i can't find the session Id.
```
tidb> show processlist;
+------+------+---------------+------+---------+------+-------+------------------+
| Id   | User | Host          | db   | Command | Time | State | Info             |
+------+------+---------------+------+---------+------+-------+------------------+
|    2 | root | 127.0.0.1     | test | Query   |    0 | 2     | show processlist |
|    4 | root | 192.168.10.72 | test | Query   |    0 | 2     |                  |
+------+------+---------------+------+---------+------+-------+------------------+
2 rows in set (0.00 sec)

```

4. What version of TiDB are you using (`tidb-server -V`)?
```
tidb> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                           |
+----------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v1.0.1-1-gaee0b39
Git Commit Hash: aee0b39af29c87742811f02c657dcf2b0fe1d60c
Git Branch: release-1.0
UTC Build Time: 2017-11-01 01:51:09 |
+----------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```

### PR DESCRIPTION
**Pull request by zimulala**
If we update handle failed, that makes us need to redo the previous operation of adding index.
Encounter an error like "update handle err [tikv:9001]PD server timeout[try again later]" in #4986.
Fix #4986

### CODE REVIEW

File: ddl/reorg.go [hunk 1]
```diff
@@ -27,6 +27,20 @@ import (
 	"github.com/pingcap/tidb/util/mock"
 )
 
+type reorg struct {
+	// doneCh is for reorganization, if the reorganization job is done,
```
Comment (coocood) [inline ddl/reorg.go]: We can remove `is for reorganization` in the comments.

File: ddl/reorg.go [hunk 2]
```diff
@@ -27,6 +27,20 @@ import (
 	"github.com/pingcap/tidb/util/mock"
 )
 
+type reorg struct {
```
Comment (coocood) [inline ddl/reorg.go]: how about `reorgCtx`?

**Comment by coocood** [comment]
I think the title of this PR should be `save reorg doneHandle regularly`.
The old code only safe doneHandle when an error occurred, but the cause of that error may also cause update done handle to fail.

**Comment by zimulala** [comment]
PTAL @coocood @shenli @winkyao @breeswish

File: ddl/reorg.go [hunk 3]
```diff
@@ -27,6 +27,21 @@ import (
...
 )
 
+// reorgCtx is for reorganization.
+type reorgCtx struct {
+	// doneCh is used to notify.
+	// If the reorganization job is done, we will use this channel to notify outer.
+	// TODO: Now we use goroutine to simulate reorganization jobs, later we may
+	// use a persistent job list.
+	doneCh chan error
+	// rowCount uses to simulate a job's row count.
```
Comment (coocood) [inline ddl/reorg.go]: is used to

File: ddl/reorg.go [hunk 4]
```diff
@@ -27,6 +27,21 @@ import (
...
+	rowCount int64
+	// notifyCancelReorgJob is used to notify the backfilling goroutine if the DDL job is cancelled.
+	notifyCancelReorgJob chan struct{}
+	// doneHandle uses to simulate the handle that is processed.
```
Comment (coocood) [inline ddl/reorg.go]: ditto

**Comment by zimulala** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

File: ddl/reorg.go [hunk 5]
```diff
@@ -27,6 +27,21 @@ import (
...
+	// doneCh is used to notify.
+	// If the reorganization job is done, we will use this channel to notify outer.
+	// TODO: Now we use goroutine to simulate reorganization jobs, later we may
+	// use a persistent job list.
+	doneCh chan error
+	// rowCount is used to simulate a job's row count.
+	rowCount int64
+	// notifyCancelReorgJob is used to notify the backfilling goroutine if the DDL job is cancelled.
+	notifyCancelReorgJob chan struct{}
+	// doneHandle is used to simulate the handle that is processed.
```
Comment (hanfei1991) [inline ddl/reorg.go]: has been processed

File: ddl/reorg_test.go [hunk 6]
```diff
@@ -62,37 +62,56 @@ func (s *testDDLSuite) TestReorg(c *C) {
...
 	for i := 0; i < 1000; i++ {
 		time.Sleep(5 * time.Millisecond)
-		err = d.runReorgJob(job, f)
+		err = d.runReorgJob(m, job, f)
 		if err == nil {
 			c.Assert(job.RowCount, Equals, rowCount)
-			c.Assert(d.reorgRowCount, Equals, int64(0))
+			c.Assert(d.reorgCtx.rowCount, Equals, int64(0))
+
+			err = ctx.Txn().Commit(goctx.Background())
```
Comment (shenli) [inline ddl/reorg_test.go]: Add comment for the purpose of this test case.

**Comment by zimulala** [comment]
PTAL @shenli @hanfei1991

**Comment by shenli** [comment]
LGTM

**Pull request merged**